src/components/List.js

```diff
@@ -3,10 +3,23 @@ import useFirestore from '../hooks/useFirestore';
 
 const List = ({ token }) => {
   const { docs } = useFirestore(token);
+
   return (
     <div>
       <h1>List</h1>
-      <ul>{docs && docs.map((doc) => <li key={doc.id}>{doc.itemName}</li>)}</ul>
+      {docs.length === 0 && (
+        <section>
+          <p>Your shopping list is currently empty.</p>
+          <a href="/add-item">Add an Item</a>
+        </section>
+      )}
+      <ul>
```

Comment:
Rachel and I tried two different ways of adding the `Loading...` UI, here is the other:
```
 {docs ? <ul>{docs.map((doc) => <li key={doc.id}>{doc.itemName}</li>)}</ul> : <div>Loading...</div>}
```
are there any opinions on which one might be a better approach? Also, the docs load so quickly for us that it was difficult to confirm that the `Loading...` is what shows up instead of the `"your shopping list is empty"` message. Are there any suggestions on how to confirm the Loading message is the actual element that's appearing.

Reply:
With regards to checking whether this is working, you can try throttling the browser. Go to the network tab and click on the dropdown that says "online" 
![image](https://user-images.githubusercontent.com/34964885/106321980-fc3d7780-6229-11eb-932f-0c56a857608c.png)

I believe that the loading message may actually not be showing at all right now because we are checking for the truthy/falsy-ness of docs, which is always an array is always truthy.

What I think we'd want to do is add a loading state to our `useFirestore` hook, which we would set to true just before we call the firbase db and set to false just after. Then, we can have `useFirestore` return that loading state in the object that is currently returning docs. 

This would actually make our hook interface act a lot like really popular data fetching hooks like react-query (https://react-query.tanstack.com) and apollo client's useQuery hook (https://www.apollographql.com/docs/react/data/queries/). :-)

Reply:
One more thing to consider, is that if we're monkeying around with the useFirestore hook, it may be a good idea to add a try/catch block around the call to the database and think about how we'd like the app to behave should there be an error returned from firebase. 

Any time we make an API call/are relying on an external dependency, it's always a good idea to think about what would happen if our "happy path" got a little less happy and the external dependency started acting up. Maybe we could do something like show a snackbar message (https://www.npmjs.com/package/react-simple-snackbar).

Reply:
Okay! So Rachel and I worked on this a bit. We tried adding a try catch and an error. Here is our `try catch` code in the `useFirestore` hook : 

<img width="520" alt="useFirestore with try catch" src="https://user-images.githubusercontent.com/36529777/106340281-1721e300-624e-11eb-8d21-6d92ea57dd15.PNG">

And the code in the `List` component : 

<img width="500" alt="List with error and loading" src="https://user-images.githubusercontent.com/36529777/106340321-3ae52900-624e-11eb-9a38-fd84696c668a.PNG">

The issue we encountered : both the `your list is empty` and the `loading` message would appear at the same time after adding an item.

Without the `try-catch` we no longer have that issue, but now when you come to your list, even if it 
has grocery items, you'll see the `your list is empty` message until the docs load. Here's the `useFirestore` hook for that : 

<img width="326" alt="useFirestore no try catch" src="https://user-images.githubusercontent.com/36529777/106340847-fa86aa80-624f-11eb-8ee6-5948cf70325c.PNG">

We like the idea of using a snackbar message or alert message, but wanted to check with @DeeDowns and @veeteeran to see what they think.

Reply:
I believe that the issue may have to do with trying to perform the cleanup from inside of the `try` block. You may want to try either reassigning the `unsubscribe` that is declared outside of the `try` block or you could extract the data fetching into another function and return `unsubscribe`. I'll post examples in a sec, but I have to jump offline for a little bit.

Reply:
Example returning unsubscribe from a callback: 
![image](https://user-images.githubusercontent.com/34964885/106343230-33774d00-6259-11eb-9a38-671637f3e6bb.png)

Example with reassignment of unsubscribe:
![image](https://user-images.githubusercontent.com/34964885/106343277-71747100-6259-11eb-93ba-cafc18c89817.png)

Reply:
Thanks for the suggestions @MikeBLambert . We tried both and unfortunately the issue is stubborn and neither resolved the problem, so we focused on just adding the `try catch`
 and made a comment to help remind us to tackle the loading issue at a later point in time.

Here is an image of the error handling we added : 

<img width="217" alt="error UI for list view" src="https://user-images.githubusercontent.com/36529777/106371761-27ed5a00-631d-11eb-9c31-d3c22025ad39.PNG">

Reply:
That is my favorite error message ever. 🤣